File: src/dotnet/Common/Models/ResourceProviders/Configuration/APIEndpointConfiguration.cs
using FoundationaLLM.Common.Constants.Authentication;
using FoundationaLLM.Common.Constants.Configuration;
using FoundationaLLM.Common.Constants.ResourceProviders;
using System.Text.Json.Serialization;

namespace FoundationaLLM.Common.Models.ResourceProviders.Configuration
{
    /// <summary>
    /// Provides the configuration for an API endpoint resource.
    /// </summary>
    public class APIEndpointConfiguration : ResourceBase
    {
        /// <summary>
        /// Creates a new instance of <see cref="APIEndpointConfiguration"/>.
        /// </summary>
        public APIEndpointConfiguration() =>
            Type = ConfigurationTypes.APIEndpointConfiguration;

        /// <summary>
        /// The api endpoint category.
        /// </summary>
        [JsonPropertyName("category")]
        [JsonConverter(typeof(JsonStringEnumConverter))]
        public required APIEndpointCategory Category { get; set; }

        /// <summary>
        /// The type of authentication required for accessing the API.
        /// </summary>
        [JsonPropertyName("authentication_type")]
        [JsonConverter(typeof(JsonStringEnumConverter))]
        public required AuthenticationTypes AuthenticationType { get; set; }

        /// <summary>
        /// The base URL of the API endpoint.
        /// </summary>
        [JsonPropertyName("url")]
        public required string Url { get; set; }

        /// <summary>
        /// The status URL of the API endpoint, if available.
        /// </summary>
        [JsonPropertyName("status_url")]
        public string? StatusUrl { get; set; }

        /// <summary>
        /// A list of URL exceptions.
        /// </summary>
        [JsonPropertyName("url_exceptions")]
        public List<UrlException> UrlExceptions { get; set; } = [];

        /// <summary>
        /// Dictionary with values used for authentication.
        /// <para>
        /// For the list of supported keys, see <see cref="AuthenticationParametersKeys"/>.
        /// </para>
        /// </summary>
        [JsonPropertyName("authentication_parameters")]
        public Dictionary<string, object> AuthenticationParameters { get; set; } = [];

        /// <summary>
        /// The timeout duration in seconds for API calls.
        /// </summary>
        [JsonPropertyName("timeout_seconds")]
        public required int TimeoutSeconds { get; set; }

        /// <summary>
        /// The name of the retry strategy.
        /// </summary>
        [JsonPropertyName("retry_strategy_name")]
        public required string RetryStrategyName { get; set; }

        /// <summary>
        /// The API endpoint provider.
        /// <para>
        /// For a list of available API endpoint providers, see <see cref="APIEndpointProviders"/>.
        /// </para>
        /// </summary>
        [JsonPropertyName("provider")]
        public string? Provider { get; set; }

        /// <summary>
        /// The version to use when calling the API represented by the endpoint.
        /// </summary>
        [JsonPropertyName("api_version")]
        public string? APIVersion { get; set; }

        /// <summary>
        /// Type of operation the endpoint is performing.
        /// This value should be completions or chat.
        /// Default value is chat.
        /// </summary>
        [JsonPropertyName("operation_type")]
        public string? OperationType { get; set; }

    }

    /// <summary>
    /// Represents an exception to the base URL.
    /// </summary>
    public class UrlException
    {
        /// <summary>
        /// The user principal name.
        /// </summary>
        [JsonPropertyName("user_principal_name")]
        public required string UserPrincipalName { get; set; }

        /// <summary>
        /// The alternative URL.
        /// </summary>
        [JsonPropertyName("url")]
        public required string Url { get; set; }

        /// <summary>
        /// Indicates whether the exception is enabled.
        /// </summary>
        [JsonPropertyName("enabled")]
        public required bool Enabled { get; set; }
    }
}

File: src/python/PythonSDK/foundationallm/models/resource_providers/configuration/api_endpoint_configuration.py
from pydantic import Field
from typing import List, Optional
from foundationallm.models.resource_providers import ResourceBase
from foundationallm.models.resource_providers.configuration import (
    ConfigurationTypes,
    UrlException
)

class APIEndpointConfiguration(ResourceBase):
    """
    API Endpoint Configuration model.
    """
    type: str = Field(default=ConfigurationTypes.API_ENDPOINT, description="The type of the API endpoint configuration.")
    category: str = Field(description="The category of the API endpoint configuration.")
    authentication_type: str = Field(description="The type of authentication required for accessing the API endpoint.")
    url: str = Field(description="The base URL of the API endpoint.")
    url_exceptions: List[UrlException] = Field(default={}, description="List of URL exceptions for the API endpoint.")
    authentication_parameters: dict = Field(default={}, description="Dictionary with values used for authentication.")
    timeout_seconds: int = Field(default=60, description="The timeout duration in seconds for API calls.")
    retry_strategy_name: str = Field(description="The name of the retry strategy to use for API calls.")
    provider: Optional[str] = Field(description="The provider of the API endpoint.")
    api_version: Optional[str] = Field(description="The version to use when calling the API represented by the endpoint.")
    operation_type: Optional[str] = Field(description="The type of operation the API endpoint is performing.")
